### PR TITLE
[PRETEST] Technical proposal for the checksum check of the three-party dependencies

### PR DESCRIPTION
**Pull request by varshith257**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references:

- [x] N
- [ ] Y 

#### 2. What is the scope of this PR (e.g. component or file name):

docs/proposal

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [x] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [ ] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

### CODE REVIEW

**Comment by varshith257** [comment]
Please have a look

cc: @zong-zhe @Peefy @AkashKumar7902

File: docs/proposal/checksum_verification_proposal.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    Checksum      string            `json:"checksum" toml:"checksum,omitempty"`
+    downloader.Source `json:"-"`
+}
+```
+
+Currently most of the code for `checksum verification` is done in `pkg/modfile.go`, `pkg/package.go` and at `client/client.go`. The current implementation for checksum is handled by `GenCheckSum` method and `GenOciManifestFromPkg` is generating an OCI manifest with name,version,desc, sum details. In `client.go` from `AcquireDepSum` method, we are fetching dependency from OCI and verifying the Checksum. This mostly setups the part of it and we have also maintaining metadata of dependencies in `kcl.mod.lock` including checksums.
+
+To extend the checksum verification, I think we can include the checksum verification/validation steps in `LoadPkgFromPath`, `LoadModFile`, `LoadLockDeps` and `ResolvePkgDepsMetadata` functions which involves the loading of package info including fetching checksums from OCI, verifying them during the loading and updating of dependencies. Also we can extend it to `Updatedeps` function to verify checksums as part of update process to ensure any modifications to dependencies trigger reverification of checksums.
```
Comment (zong-zhe) [inline docs/proposal/checksum_verification_proposal.md:85]: Performance degradation must be considered, as fetching checksums are dependent on network conditions and can have a significant performance impact.

Comment (varshith257) [reply docs/proposal/checksum_verification_proposal.md:85]: Noted!

**Comment by varshith257** [comment]
@zong-zhe Can we look ahead of database design now?

**Comment by zong-zhe** [comment]
> @zong-zhe Can we look ahead of database design now?

Hi @varshith257 , I would like to provide you with the current status of KCL:

Currently, KCL's primary third-party dependencies are stored through the OCI Registry. The KCL packages are also packaged into OCI Artifacts, which provide an OCI Manifest to store informations. In previous work, KPM has also included some fields in the annotation block of OCI manifest.

Add some field in OCI manifest when push: https://github.com/kcl-lang/kpm/blob/ecbd431ede82867f9bc29aed5f4e1ea20da964d8/pkg/client/client.go#L1598

The main concern we need to address is the timing of the checksum requests, as this process can introduce network delays. KCL has a high-performance requirement for the compilation process, so we need to carefully consider when to make network requests during the checksum verification and provide an offline mode or cache to improve performance.

**Comment by zong-zhe** [comment]
Hi @varshith257 😀

In addition, you can try to design a separate module called `Checker`, which can be used to check whether a KCL dependency is valid. This includes checking whether the name of the dependency is appropriate, whether the version number matches the semantic version, and the checksum, rather than simply making changes to an existing process.

**Comment by varshith257** [comment]
> Hi @varshith257 😀
> 
> In addition, you can try to design a separate module called `Checker`, which can be used to check whether a KCL dependency is valid. This includes checking whether the name of the dependency is appropriate, whether the version number matches the semantic version, and the checksum, rather than simply making changes to an existing process.
> 
> 

Makes sense of having a separate module as Checker.

**Pull request closed without merging**